### ISSUE TITLE
Observer with render prop and inject

### ISSUE THREAD

**Issue by FredyC** [mobxjs/mobx-react#387]
I would like to propose an alternative way of using `Observer`. There is a clear benefit, that you don't need to repeat closing element there and it feels bit cleaner in my opinion.

```js
<Observer
  render={() => (
    <div />
  )}
/>
// vs (note: both are formatted by Prettier)
<Observer>
  {() => (
    <div />
  )}
</Observer>
```

Furthermore, it's a bit of pain needing to use `inject` HOC with this pattern. It separates two concerns that should stick together in my opinion. What about something like this?

```js
<Observer inject={[ 'store1', 'store2' ]} render={({ store1, store2 }) => (
  <div>{store1.observable && store2.observable}</div>
)} />
```

I am not entirely sure it's possible to achieve this with the current implementation of the `Observer` which is just wrapped HOC under the hood. Perhaps the whole thing could go other way around, implement `Observer` as the regular class with everything and then turning it to HOC is much easier.

**Comment by mweststrate** [comment]
@FredyC that looks nice indeed. I think something like that could quite easily be implemented as just sugar around the current api's. Feel free to do so!

**Comment by mweststrate** [comment]
Released in 4.4.0

**Comment by urugator** [comment]
I think that introducion of `inject` prop on `Observer` wasn't such a good idea as it can make `props` handling quite tricky:
```javascript
const MyCmp = props => <Observer 
  inject={(stores, props) => ({ store: stores.store, prop: props.prop })}
  render={props => (
    <div>
      props.store.something + props.prop      
    </div> 
  )}
/>
```
There are 3 different props objects:
Props from surrounding context (in this case props accepted by `MyCmp`)
Props passed to `inject` mapper, which are always empty...
Props passed to `render` which are the props returned from `inject` mapper...

Basically the problem is that `Observer` doesn't accept any props on it's own, therefore `inject` mapper has no props to intercept. Therefore I think the injection should be limited to "static" injection or removed altogether.

See #419

**Comment by urugator** [comment]
@FredyC Seems it tricked you as well ... have you noticed that `prop` is always `undefined`?

**Comment by FredyC** [comment]
Well, that's a different problem that `inject` attribute is not exposing props in the second argument and I don't even think it's needed. This role is well suited by eg. `withProps` from `recompose` library. I wouldn't be dragging it here. Yea I know that's HoC world so it's kinda nicer in here. But if anything, I would introduce different prop for mapping component props then mixing these two concerns together.

I was merely pointing out here that there doesn't need to be any confusion with `props` variable as long as you use destructuring.

**Comment by urugator** [comment]
> use destructuring

The naming isn't such an issue, but it's not immeditately clear how props are (not) propagated.

> inject attribute is not exposing props in the second argument

Have fun explaining differences between `inject()` and `inject=` to users.

Shouldn't `render` accept `stores` object directly instead of `props` (which are not actually props, but only filtered stores)?:
```javascript
const MyCmp = props => <Observer   
  render={({ store }) => ( 
    <div>
      store.something + props.something     
    </div> 
  )}
/>
```
How is it useful to have the store mapping outside of `render` callback? I think it doesn't make sense semantically or practically.

**Comment by FredyC** [comment]
@urugator Can you show me where in the docs is a mention that `inject` provides access to props? I think you are bit confused. What `inject` gets is not any sort of filtered component props, but props which you pass into `Provider` component. It's correct to call this "stores" as it's not really any kind of props in that scope anymore.

Injecting all stores into `render` prop does not sound right because you don't want to inject every time. By omitting `inject` prop, the `inject` HoC won't be used at all which is imo the desired effect. It's useful to have store mapping separated so you can define a bunch of different selector functions that can be reused. 

By the way, if you want to achieve the behavior you have mentioned, simple identity function is enough. If you like you can even write your own wrapper Observer that always injects everything, but imo that's a bad practice.

```jsx
const Injector = ({ render }) => <Observer
  inject={(stores) => stores}
  render={(stores) => render(stores)}
/>
```

**Comment by urugator** [comment]
The second arg: `(allStores, nextProps, nextContext) => additionalProps`
https://github.com/mobxjs/mobx-react#with-typescript

Reusability and different mappers:
```javascript
const MyCmp = props => <Observer   
  render={({ store }) => <Reusable prop={store} />}
/>

// Different mappers:
const map1 = ({ store1 })  => ({ prop: store1 });
const map2 = ({ store2 })  => ({ prop: store2 });

const MyCmp1 = props => <Observer   
  render={stores => <Reusable ...map1(stores) />}
/>

const MyCmp2 = props => <Observer   
  render={stores => <Reusable ...map2(stores) />}
/>
```

**Comment by FredyC** [comment]
Yea, I've edited my answer as I at first did not spot that signature for `inject`, but as I am saying there is no reason to bother with `nextProps` with this pattern at all. And once again, it's necessary to have `inject` prop there so an actual injecting is not happening when you don't need it.

**Comment by urugator** [comment]
> it's the same thing you are getting in the top level component. // nextProps === myProps

> I am not sure why you are confused about it.

This is golden :)

**Comment by FredyC** [comment]
Ok, I see now where are you coming at you. You are right, a `nextProps === myProps` is not correct. However, I stand by the fact, that you don't need to worry about `nextProps` as there is nothing you can use in this case.

Either way I am against magically injecting all stores to `render` because its **Observer** which means it should be observing things and not injecting as some kind of side effect. Consider this example.

```jsx
// come to think of this, would be lovely to have a render prop component for injecting only
const decorate = inject(({ userStore }) => ({ userStore }))

const ParentComponent = decorate(({ userStore }) => (
  <div>
    <FirstChild userStore={userStore} />
    <SecondChild userStore={userStore} />
    <ThirdChild userStore={userStore} />
  </div>
)

const FirstChild = ({ userStore }) => <Observer
  render={() => <div>{userStore.username}</div>}
/>
// similarly for SecondChild and ThirdChild using different userStore values/actions
```

As you can see, Observer is used for _observing_ purposes only and it does not need to inject any stores. This support separation of concerns and is especially useful in tests as I can pass `userStore` object only without need of wrapping with `Provider`.

So yea, I am definitely against auto-injecting. However I can live with `inject` prop being a boolean (false by default) that inject all stores. That might work well for both our cases.

**Comment by urugator** [comment]
> you don't need to worry about nextProps as there is nothing you can use in this case.

That's exactly the reason why I worry. There is something which is useless and doesn't make sense. That's obviously potential source of confusion and sign of something wrong with the design.

> Observer is used for observing purposes only and it does not need to inject any stores.
> I am definitely against auto-injecting

Nobody forces you to access stores inside render, exactly the same way as nobody forces you to use or access stores in inject.
`Observer` always provides all stores ("auto-injects everything"), it doesn't matter if it does so for one prop (`inject`) or another (`render`) as both of these props are defined at the same place.
If you move inject to render callback nothing really changes in terms of coupling, just the syntax: 

```javascript
<Observer
  inject="inject here"
  render="render here"
/>

<Observer  
  render={
    "inject here"
    "render here"
  }
/>

<Observer // observer with inject only 
  render={() => {
    "inject here"
    return <Observer // observer with render only
      render="render here"
    />  
  }}
/>
```  
 
> This support separation of concerns

*inject prop* doesn't provide any advatages in terms of SoC: 

```javascript
// Concerns not separeted (together in Observer)
// with inject prop
inject={({ userStore }) => ({ userStore })} // map
render={({ userStore }) => { // render
  return <div>{userStore.username}</div>
}}
// without inject prop
render={({ userStore }) => {
  const {({ userStore }) = ({ userStore })}; // map
  return <div>{userStore.username}</div> // render 
}}

// Separating concerns (concerns implemented outside and composed in Observer)
const map = ({ userStore }) => ({ userStore })
const render = ({ userStore }) => <div>{userStore.username}</div>
// with inject prop
inject=({ userStore }) => map({ userStore })
render=({ userStore }) => render({ userStore })
// without inject prop
render=({ userStores }) => {
  ({ userStore }) => map({ userStore }) 
  return render({ userStore })
}}
```
The inject and render are always together inside `Observer`, they're not separated, until you actually separate their logic, but you don't need separate *inject prop* to do that.  
The only difference is how things are composed together. With *inject prop*, mapper and render are composed as props. Without *inject prop*, mapper and render are composed as standard functions.  
Without *inject prop* it's imho easier to follow how data are propagated, because data don't have to be passed horizontally from prop (`inject`) to prop (`render`) on the same level, but naturally from top to bottom as they're scoped.
Composing logic via "props" is a misuse as it's not obvious how individual props (and therefore the logic) relates to one another. 

> Consider this example.

Can you demonstrate the usage of *inject prop* in your example please? I would show you how to do the same without *inject prop* and with the same amount of SoC/testability/reusability. (if it's still not clear the examples)
Your example in it's current form would stay exactly as it is ... again ... nothing forces you to access stores in `FirstChild`'s' `render`, exactly the same way as nothing forces you to introduce *inject prop* and do exactly the same thing there.
`FirstChild` either depends on `stores` or it doesn't ... it doesn't matter if depednecy on `stores` is inside `render` prop or inside `inject` prop as both of these props are always defined on the same place and relates to the same component (`Observer`).   

> you see that injecting is not especially cheap operation

This is because normally `inject` has to deal with reactivity and props. But in this case, there are no props and render is already reactive.
Therefore `Observer` only has to expose `context.mobxStores` to `render` and that's basically it ... that should be actually lot cheaper than current impl ...

**Comment by FredyC** [comment]
Sorry, I haven't read everything, too long and I don't care that much :) You are still missing the point obviously. Please check the source code [on this line](https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L357). Think about it and tell me how that makes this statement of yours truthy :)

> Observer always provides all stores ("auto-injects everything")

Obviously, this is not true. Also same applies to `@observer` which lacks any kind of injecting so I don't see why `Observer` should inject by default. You want to introduce inconsistency here ;)

**Comment by urugator** [comment]
"Observer always **provides** all stores" - meaning that all stores are always available to the content of `Observer` and you're always in control which stores you actually "inject".
There is no practical difference between grabbing(injecting) the relevant stores inside "inject" or inside "render". In both cases you have exactly the same amount of control of what you can use (what is provided) and what you actually use (what you inject). Splitting this into two different props is absolutely artifical and doesn't provide any real benefits (soc/testability/performance/composition) as demonstrated in previous comments.  
How is this actually implemented is unrelated from the usage perspective. It could be easily implemented without the mentioned condition, always wrapping the component inside Injector, without actually passing anything to "render". It wouldn't be optimal performance wise, but it's an implementation detail and as I mentioned in last paragraph the actual implmentation should be a lot simpler.

> You want to introduce inconsistency here ;)

The "inconsistency" was introduced by providing `inject` to `Observer`.
If I would apply your logic that `Observer` doesn't use `inject` by default, then `@observer` should optionally accept `inject` as well to make it consistent with `Observer`.
Btw `observer` actually accepted inject as a first param in previous versions. It was removed as it was kinda redundant: `observer(opts, cmp)` VS `inject(opts)(observer(cmp))`.

**Comment by FredyC** [comment]
I give up. If someone else wants to chime in here, please do. I am unable to answer in the way it make sense obviously :)

**Comment by urugator** [comment]
Btw I think that current implmentation is actually broken, see #423

**Comment by mweststrate** [comment]
Only scanned above conversation superficially, but if I am not mistaken, we could make the `inject` work properly if we limited support only to the accepting one or multiple store _names_. correctly? 

I'm not really in favor or against automatically injecting all the stores from a conceptual point of view. But practically it means that we don't know if stores are actually injected. Subscribing (in future: wrapping) to an unused contextType is undesirable I think. Yet we could detect that `arguments.length > 0` and only subscribe to the stores in that case 

Tl;DR: So we could I think take one of those two approaches
1. only support store names: `<Observer inject="bookstore" render=({ bookStore }) => {} />`
2. inject anything, of render has `1` argument: `<Observer render=({ bookStore }) => {} />`

I think I agree with @urugator that supporting `inject` functions separately doesn't really make that much sense. If you really need that, just refactor your code a bit and extract it to a separate component.

**Comment by urugator** [comment]
@mweststrate

> Subscribing (in future: wrapping) to an unused contextType is undesirable I think.

Do we need to know precisely which stores are being injected or just the fact that some (any) store is being injected?
If it's the latter I propose to remove injection from `<Observer>` and replace it with specific component: 
```javascript
<Inject render={({ store }) => <div>store.foo</div>} /> // the render is automatically wrapped in `<Observer />` 
```
I think it also makes more sense in relation to #410

**Comment by urugator** [comment]
> Yet we could detect that arguments.length > 0 and only subscribe to the stores in that case

Is it really possible? Note that `render` function is passed via props, therefore is dynamic and can change, but the component and cotextType must be static right? ... the situation with `inject` prop is quite the same... we can't decide the `contexType` on runtime ... or can we?

**Comment by FredyC** [comment]
> But practically it means that we don't know if stores are actually injected.

I don't agree with that because if you have `inject` prop in place, you are explicitly saying that stores will be injected. It's much more expressive and readable imo. I am not saying it needs to be a function. It can be a string or simply a boolean saying that you want all stores from the context.

By the name Observer, it's kinda clear that its main purpose is observing. Doing inject is a bonus and should be treated that way. Especially if there is no way of preventing the use of context dynamically, it should be done when you want to inject only.

**Comment by rtpm** [comment]
Like the provider with inject prop like https://github.com/jamiebuilds/unstated and subscribing to specified store via <Subscribe> (or call it Observer).
